File: ios/Classes/FlutterImageCompressPlugin.m
#import "FlutterImageCompressPlugin.h"
#import "CompressListHandler.h"
#import "CompressFileHandler.h"

BOOL showLog = false;

@implementation FlutterImageCompressPlugin
static dispatch_queue_t serial_queue;

+ (void)registerWithRegistrar:(NSObject <FlutterPluginRegistrar> *)registrar {
    serial_queue = dispatch_queue_create("com.github.flutter_compress.SerialQueue", NULL);

    FlutterMethodChannel *channel = [FlutterMethodChannel
            methodChannelWithName:@"flutter_image_compress"
                  binaryMessenger:[registrar messenger]];
    FlutterImageCompressPlugin *instance = [[FlutterImageCompressPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    dispatch_sync(serial_queue, ^{
        if ([call.method isEqualToString:@"compressWithList"]) {
            CompressListHandler *handler = [[CompressListHandler alloc] init];
            [handler handleMethodCall:call result:result];
        } else if ([@"compressWithFile" isEqualToString:call.method]) {
            CompressFileHandler *handler = [[CompressFileHandler alloc] init];
            [handler handleMethodCall:call result:result];
        } else if ([@"compressWithFileAndGetFile" isEqualToString:call.method]) {
            CompressFileHandler *handler = [[CompressFileHandler alloc] init];
            [handler handleCompressFileToFile:call result:result];
        } else if ([@"showLog" isEqualToString:call.method]) {
            [self setShowLog:[call arguments]];
            result(@1);
        } else {
            result(FlutterMethodNotImplemented);
        }
    });

}

+ (BOOL)showLog{
    return showLog;
}

- (void)setShowLog:(BOOL)log{
    showLog = log;
}

@end

File: android/src/main/kotlin/com/example/flutterimagecompress/FlutterImageCompressPlugin.kt
package com.example.flutterimagecompress

import android.os.Build
import com.example.flutterimagecompress.core.CompressFileHandler
import com.example.flutterimagecompress.core.CompressListHandler
import com.example.flutterimagecompress.format.FormatRegister
import com.example.flutterimagecompress.handle.common.CommonHandler
import com.example.flutterimagecompress.handle.heif.HeifHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

class FlutterImageCompressPlugin(private val registrar: Registrar) : MethodCallHandler {
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar): Unit {
      val channel = MethodChannel(registrar.messenger(), "flutter_image_compress")
      val flutterImageCompressPlugin = FlutterImageCompressPlugin(registrar)
      channel.setMethodCallHandler(flutterImageCompressPlugin)
    }

    var showLog = false
  }

  init {
    FormatRegister.registerFormat(CommonHandler(0)) // jpeg
    FormatRegister.registerFormat(CommonHandler(1)) // png
    FormatRegister.registerFormat(HeifHandler()) // heic / heif
    FormatRegister.registerFormat(CommonHandler(3)) // webp
  }

  override fun onMethodCall(call: MethodCall, result: Result): Unit {
    when (call.method) {
      "showLog" -> result.success(handleLog(call))
      "compressWithList" -> CompressListHandler(call, result).handle(registrar)
      "compressWithFile" -> CompressFileHandler(call, result).handle(registrar)
      "compressWithFileAndGetFile" -> CompressFileHandler(call, result).handleGetFile(registrar)
      "getSystemVersion" -> result.success(Build.VERSION.SDK_INT)
      else -> result.notImplemented()
    }
  }

  private fun handleLog(call: MethodCall): Int {
    val arg = call.arguments<Boolean>()
    showLog = (arg == true)
    return 1
  }

}
